### PR TITLE
[10.0][NEW] pos_invoice_pay module

### PR DESCRIPTION
**Pull request by ilmir-k**
[10.0][NEW] pos_invoice_pay module

### CODE REVIEW

**Comment by ilmir-k** [comment]
issue for unitest: #569

**Comment by yelizariev** [comment]
- [ ] @ArtyomLosev  `phantomjs: u"null is not an object (evaluating 'order.get_paymentlines') render_paymentlines@http://127.0.0.1:8069/pos_invoice_pay/static/src/js/main.js:867`

File: pos_invoice_pay/static/description/index.html [hunk 1]
```diff
@@ -0,0 +1,242 @@
...
+  </div>
+</section>
+
+<section class="oe_container">
+  <div class="oe_row oe_spaced">
+	<h3 class="oe_slogan">How It Works</h3>
+	<h4 class="oe_slogan"><b>Fetch Sale Orders</b></h4>
+	<div class="oe_span12 text-center">
+      <p class="oe_mt32">
+ 		<font size="3px">Go to <em>Sales &rarr; Sales Orders</em> menu. Create and confirm a sale order.</font>
```
Comment (yelizariev) [inline pos_invoice_pay/static/description/index.html]: clean up html. WHat is ``3px``?

File: pos_invoice_pay/static/description/index.html [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+    </div>
+  </div>
+</section>
+
+
+<section class="oe_container">
+  <div class="oe_row oe_spaced">
+    <div class="oe_span12">
+      <p class="oe_mt32">
+		<div class="alert alert-info" style="padding:8px;font-weight: 300; font-size: 20px;">
```
Comment (yelizariev) [inline pos_invoice_pay/static/description/index.html]: Clean up html https://odoo-development.readthedocs.io/en/latest/description/index.html.html#key-features

File: pos_invoice_pay/static/description/index.html [hunk 3]
```diff
@@ -0,0 +1,242 @@
...
+      <img style="max-width: 80%" src="so_details.png"/>
+    </div>
+   </div>
+</section>
+
+<section class="oe_container">
+  <div class="oe_row oe_spaced">
+    <div class="oe_span12 text-center">
+      <p class="oe_mt32">
+ 		<font size="3px">Once you click on <span style="color: black;">Create Invoice</span> button, the corresponding invoice is created and validated automatically, and you will be redirected to POS payment screen. Click on <span style="color: black;">Validate</span> button to register payment. You can see Sale Order and Invoice reference on payment receipt.</font>
```
Comment (yelizariev) [inline pos_invoice_pay/static/description/index.html]: clean up html. WHat is 3px?

File: pos_invoice_pay/static/description/index.html [hunk 4]
```diff
@@ -0,0 +1,242 @@
...
+      </p>		
+    </div>
+    <div class="oe_screenshot" align="center">
+      <img style="max-width: 80%" src="pay_so_screen.png"/>
```
Comment (yelizariev) [inline pos_invoice_pay/static/description/index.html]: Use existing classes 
https://odoo-development.readthedocs.io/en/latest/description/index.html.html#text-image

**Review by yelizariev — CHANGES_REQUESTED** [review]
Clean up html and add clean templates to odoo-development if you use new template

**Comment by ilmir-k** [comment]
it's strangely that `--force-with-lease` didn't affect

**Comment by yelizariev** [comment]
``--force-with-lease`` works exactly as it's should be

File: pos_invoice_pay/models.py [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+            inv_obj = self.env['account.invoice'].search([('id', '=', inv_id)])
+            journal_id = statement[2]['journal_id']
+            journal = self.env['account.journal'].search([('id', '=', journal_id)])
+            amount = statement[2]['amount']
+            cashier = invoice['data']['user_id']
+            writeoff_acc_id = False
+            payment_difference_handling = 'open'
+
+            if amount > inv_obj.residual:
+                writeoff_acc_id = self.env['account.account'].search([('code', '=', 220000)]).id
```
Comment (blaggacao) [inline pos_invoice_pay/models.py]: hard coded?

File: pos_invoice_pay/static/src/js/main.js [hunk 6]
```diff
@@ -0,0 +1,1050 @@
...
+var QWeb = core.qweb;
+var _t = core._t;
+var round_pr = utils.round_precision;
+
+
+var _super_posmodel = models.PosModel.prototype;
+models.PosModel = models.PosModel.extend({
+    initialize: function (session, attributes) {
+        var self = this;
+        this.models.push(
```
Comment (blaggacao) [inline pos_invoice_pay/static/src/js/main.js]: Nooo :smile: 

There is an official hook for that:

```js
// Loads openerp models at the point of sale startup.
// load_models take an array of model loader declarations.
// - The models will be loaded in the array order. 
// - If no openerp model name is provided, no server data
//   will be loaded, but the system can be used to preprocess
//   data before load.
// - loader arguments can be functions that return a dynamic
//   value. The function takes the PosModel as the first argument
//   and a temporary object that is shared by all models, and can
//   be used to store transient information between model loads.
// - There is no dependency management. The models must be loaded
//   in the right order. Newly added models are loaded at the end
//   but the after / before options can be used to load directly
//   before / after another model.
//
// models: [{
//  model: [string] the name of the openerp model to load.
//  label: [string] The label displayed during load.
//  fields: [[string]|function] the list of fields to be loaded. 
//          Empty Array / Null loads all fields.
//  order:  [[string]|function] the models will be ordered by 
//          the provided fields
//  domain: [domain|function] the domain that determines what
//          models need to be loaded. Null loads everything
//  ids:    [[id]|function] the id list of the models that must
//          be loaded. Overrides domain.
//  context: [Dict|function] the openerp context for the model read
//  condition: [function] do not load the models if it evaluates to
//             false.
//  loaded: [function(self,model)] this function is called once the 
//          models have been loaded, with the data as second argument
//          if the function returns a deferred, the next model will
//          wait until it resolves before loading.
// }]
//
// options:
//   before: [string] The model will be loaded before the named models
//           (applies to both model name and label)
//   after:  [string] The model will be loaded after the (last loaded)
//           named model. (applies to both model name and label)
//
exports.load_models = function(models,options) {
    options = options || {};
    if (!(models instanceof Array)) {
        models = [models];
    }

    var pmodels = exports.PosModel.prototype.models;
    var index = pmodels.length;
    if (options.before) {
        for (var i = 0; i < pmodels.length; i++) {
            if (    pmodels[i].model === options.before ||
                    pmodels[i].label === options.before ){
                index = i;
                break;
            }
        }
    } else if (options.after) {
        for (var i = 0; i < pmodels.length; i++) {
            if (    pmodels[i].model === options.after ||
                    pmodels[i].label === options.after ){
                index = i + 1;
            }
        }
    }
    pmodels.splice.apply(pmodels,[index,0].concat(models));
};
```

**Review by blaggacao — CHANGES_REQUESTED** [review]
Hi, I just had a look on the code, here is what I spotted...

This image: pos_invoice_pay/static/description/pay_invoice_screen.png Shoud there really be the invoice button available?

**Comment by yelizariev** [comment]
@blaggacao Thank you for review!

@GabbasovDinar will work on it

@ilmir-k , can you check question with pos_invoice_pay/static/description/pay_invoice_screen.png

File: pos_invoice_pay/models.py [hunk 7]
```diff
@@ -29,7 +29,7 @@ def process_invoice_payment(self, invoice):
             payment_difference_handling = 'open'
 
             if amount > inv_obj.residual:
-                writeoff_acc_id = self.env['account.account'].search([('code', '=', 220000)]).id
+                writeoff_acc_id = self.env['pos.config'].search([]).pos_invoice_pay_writeoff_account_id.id
```
Comment (GabbasovDinar) [inline pos_invoice_pay/models.py]: @yelizariev @ArtyomLosev  What happens if a query returns multiple objects? e.g. returns the pos.config(1,2,3)
I think that we will have an error

Comment (yelizariev) [reply pos_invoice_pay/models.py]: Yes

**Review by yelizariev — COMMENTED** [review]
Not blocking from my side, but functional review is required

**Review by yelizariev — APPROVED** [review]
Approved as I did earlier.

**Comment by ilmir-k** [comment]
@blaggacao 
> This image: pos_invoice_pay/static/description/pay_invoice_screen.png Shoud there really be the invoice button available?

I think we should leave it there, because someone might want to print invoice directly from POS payment screen rather than from backend

**Review by Railia — APPROVED** [review]

File: pos_invoice_pay/static/description/index.html [hunk 8]
```diff
@@ -0,0 +1,271 @@
...
+	      <i class="fa fa-check-square-o text-primary"></i>
+	      easy to pay SO&amp;Invoices over POS
+	    </li>
+	    <li>
+	      <i class="fa fa-check-square-o text-primary"></i>
+	      see SO&amp;Invoice details in corresponding menu
+	    </li>
+	    <li>
+	      <i class="fa fa-check-square-o text-primary"></i>
+	      automatically filter all SO&amp;Invoices by customer selected
```
Comment (yelizariev) [inline pos_invoice_pay/static/description/index.html]: Use spaces `` SO &amp; Invoices ``

File: pos_invoice_pay/static/description/index.html [hunk 9]
```diff
@@ -0,0 +1,271 @@
...
+	</div>
+      </p>		
+    </div>
+  </div>
+</section>
+
+<section class="oe_container">
+  <div class="oe_row oe_spaced">
+    <h3 class="oe_slogan">How It Works</h3>
+    <h4 class="oe_slogan"><b>Fetch Sale Orders</b></h4>
```
Comment (yelizariev) [inline pos_invoice_pay/static/description/index.html]: If we use bold font for subsection, update it in templates: https://odoo-development.readthedocs.io/en/latest/description/index.html.html#subsection

Comment (yelizariev) [reply pos_invoice_pay/static/description/index.html]: I think subsection can be just ``Sale Orders``

File: pos_invoice_pay/static/description/index.html [hunk 10]
```diff
@@ -0,0 +1,271 @@
...
+      <img style="max-width: 80%" src="backend_invoice.png"/>
+    </div>
+  </div>
+</section>
+
+<br/>
+
+<section class="oe_container">
+  <div class="oe_row oe_spaced">
+    <h4 class="oe_slogan"><b>Fetch Invoices</b></h4>
```
Comment (yelizariev) [inline pos_invoice_pay/static/description/index.html]: Remove or replace word Fetch

File: pos_invoice_pay/static/description/index.html [hunk 11]
```diff
@@ -0,0 +1,271 @@
...
+      <img style="max-width: 80%" src="sale_order.png"/>
+    </div>
+  </div>
+</section>
+
+<section class="oe_container">
+  <div class="oe_row oe_spaced">
+    <div class="oe_span12 text-center">
+      <p class="oe_mt32">
+ 	<font style="font-size: 120%;">Go to <em>Point of Sale &rarr; Dashboard</em> and open POS session. To see sale orders to be paid, click on <span style="color: black;">Fetch Orders</span> button</font>
```
Comment (yelizariev) [inline pos_invoice_pay/static/description/index.html]: Add to templates section *Reference to menu*
https://odoo-development.readthedocs.io/en/latest/description/index.html.html

File: pos_invoice_pay/static/description/index.html [hunk 12]
```diff
@@ -0,0 +1,271 @@
...
+  <div class="oe_row oe_spaced">
+    <h3 class="oe_slogan">How It Works</h3>
+    <h4 class="oe_slogan"><b>Fetch Sale Orders</b></h4>
+    <div class="oe_span12 text-center">
+      <p class="oe_mt32">
+ 	<font style="font-size: 120%;">Go to <em>Sales &rarr; Sales Orders</em> menu. Create and confirm a sale order.</font>
+      </p>		
+    </div>
+    <div class="oe_screenshot" align="center">
+      <img style="max-width: 80%" src="sale_order.png"/>
```
Comment (yelizariev) [inline pos_invoice_pay/static/description/index.html]: Can we remove this section? It's too boring for first picture

File: pos_invoice_pay/static/description/index.html (hunk 11 above)
Comment (yelizariev) [reply pos_invoice_pay/static/description/index.html]: No need to describe how to open POS. It's well-known and too boring

**Review by yelizariev — CHANGES_REQUESTED** [review]
Can we remove word ``Fetch`` from POS UI. Just ``Orders`` and ``Invoices``

File: pos_invoice_pay/static/description/index.html (hunk 12 above)
Comment (ilmir-k) [reply pos_invoice_pay/static/description/index.html]: I think we need to leave that. It's used for comparing, i.e. used to show invoiceable lines for SO in backend and to see how it looks in POS then

File: pos_invoice_pay/static/description/index.html (hunk 11 above)
Comment (ilmir-k) [reply pos_invoice_pay/static/description/index.html]: Done https://github.com/it-projects-llc/odoo-development/pull/152

File: pos_invoice_pay/static/description/index.html (hunk 10 above)
Comment (ilmir-k) [reply pos_invoice_pay/static/description/index.html]: Removed

File: pos_invoice_pay/static/description/index.html (hunk 9 above)
Comment (ilmir-k) [reply pos_invoice_pay/static/description/index.html]: Done https://github.com/it-projects-llc/odoo-development/pull/152

File: pos_invoice_pay/static/description/index.html (hunk 8 above)
Comment (ilmir-k) [reply pos_invoice_pay/static/description/index.html]: Done

**Comment by ilmir-k** [comment]
> Can we remove word Fetch from POS UI. Just Orders and Invoices

I renamed it as ``Sale Orders`` and ``Invoices``, because ``Orders`` may be used for POS orders history

**Comment by ilmir-k** [comment]
@yelizariev you can check it

File: pos_invoice_pay/static/description/index.html (hunk 12 above)
Comment (yelizariev) [reply pos_invoice_pay/static/description/index.html]: Move it from first position at least. E.g. after showing in POS, next slide -- picture from backend

Comment (yelizariev) [reply pos_invoice_pay/static/description/index.html]: Or screentshot at the end of story -- where you have invoice

**Comment by yelizariev** [comment]
First image still boring

File: pos_invoice_pay/static/description/index.html [hunk 13]
```diff
@@ -88,6 +82,15 @@ <h4 class="oe_slogan"><b>Sale Orders</b></h4>
     <div class="oe_screenshot" align="center">
       <img style="max-width: 80%" src="so_details.png"/>
     </div>
+    <br/>
+    <div class="oe_span12 text-center">
+      <p class="oe_mt32">
+ 	<font style="font-size: 120%;">The above sale order we get from backend</font>
```
Comment (yelizariev) [inline pos_invoice_pay/static/description/index.html]: The sale order above is gotten from backend


Sounds better?

**Review by yelizariev — APPROVED** [review]
New guidelines for copyrights are not applied. We can do it later

File: pos_invoice_pay/static/description/index.html (hunk 13 above)
Comment (ilmir-k) [reply pos_invoice_pay/static/description/index.html]: @yelizariev 
Another variant:
``The sale order above is taken from backend``

File: pos_invoice_pay/models.py [hunk 14]
```diff
@@ -0,0 +1,153 @@
...
+            writeoff_acc_id = False
+            payment_difference_handling = 'open'
+
+            if amount > inv_obj.residual:
+                writeoff_acc_id = self._default_session().config_id.pos_invoice_pay_writeoff_account_id.id
+                payment_difference_handling = 'reconcile'
+
+            vals = {
+                'journal_id': journal.id,
+                'payment_method_id': 1,
```
Comment (yelizariev) [inline pos_invoice_pay/models.py:42]: Кажется кто-то пытался нас наебать

**Pull request merged**